src/sentry/services/hybrid_cloud/rpc.py

```diff
@@ -471,89 +472,113 @@ def dispatch_remote_call(
-        raise RpcSendException("Not configured for RPC network requests")
-    path = django.urls.reverse(
-        "sentry-api-0-rpc-service",
-        kwargs={"service_name": service_name, "method_name": method_name},
-    )
-    url = address + path
-    request_body = {
-        "meta": {},  # reserved for future use
-        "args": serial_arguments,
-    }
-    data = json.dumps(request_body).encode(_RPC_CONTENT_CHARSET)
-    signature = generate_request_signature(path, data)
-    headers = {
-        "Content-Type": f"application/json; charset={_RPC_CONTENT_CHARSET}",
-        "Authorization": f"Rpcsignature {signature}",
-    }
-    timer = metrics.timer(
-        "hybrid_cloud.dispatch_rpc.duration", tags={"service": service_name, "method": method_name}
-    )
-    span = sentry_sdk.start_span(
-        op="hybrid_cloud.dispatch_rpc", description=f"rpc to {service_name}.{method_name}"
-    )
-    with span, timer:
-        if use_test_client:
-            response = _fire_test_request(path, headers, data, region)
+@dataclass(frozen=True)
+class _SiloServiceDispatch:
+    method_name: str
+    service_name: str
+    region: Region | None
+    serial_arguments: ArgumentDict
+    use_test_client: bool
+
+    def __post_init__(self) -> None:
+        if not (self.address and settings.RPC_SHARED_SECRET):
+            raise RpcSendException("Not configured for RPC network requests")
+
+    @property
+    def address(self) -> str:
+        if self.region is None:
+            return settings.SENTRY_CONTROL_ADDRESS
         else:
-            response = _fire_request(url, headers, data)
-        metrics.incr(
-            "hybrid_cloud.dispatch_rpc.response_code", tags={"status": response.status_code}
+            return self.region.address
+
+    @property
+    def path(self) -> str:
+        return django.urls.reverse(
+            "sentry-api-0-rpc-service",
+            kwargs={"service_name": self.service_name, "method_name": self.method_name},
         )
 
-        if response.status_code != 200:
-            if in_test_environment():
-                if response.status_code == 500:
-                    raise Exception(
-                        "Error invoking rpc at {path}: check error logs for more details"
-                    )
-                raise Exception("Error invoking rpc at {path}: {response.json()['detail']}")
-            # Careful not to reveal too much information in production
-            if response.status_code == 403:
-                raise Exception("Unauthorized service access")
-            if response.status_code == 400:
-                raise Exception("Invalid service request")
-            raise Exception("Service unavailable")
-        serial_response = response.json()
-    return serial_response
-
-
-def _fire_test_request(
-    path: str, headers: Mapping[str, str], data: bytes, region: Region | None
-) -> Any:
-    from django.test import Client
+    def dispatch(self) -> Any:
```

Comment:
Nit -- can we actually pass the `use_test_client` into the dispatch directly, as opposed to it being a member? It can't, and shouldn't, effect any other prep steps.  Ideally the paths between the test client and real one are identicaly up to actually dispatching.

Reply:
Yep, that works.